src/SqlAsyncCollector.cs

```diff
@@ -261,32 +281,39 @@ private static void GenerateDataQueryForMerge(TableInformation table, IEnumerabl
             // If there are duplicate primary keys, we'll need to pick the LAST (most recent) row per primary key.
             foreach (T row in rows.Reverse())
             {
-                // SQL Server allows 900 bytes per primary key, so use that as a baseline
-                var combinedPrimaryKey = new StringBuilder(900 * table.PrimaryKeys.Count());
-
-                // Look up primary key of T. Because we're going in the same order of fields every time,
-                // we can assume that if two rows with the same primary key are in the list, they will collide
-                foreach (PropertyInfo primaryKey in table.PrimaryKeys)
+                if (typeof(T) != typeof(JObject))
                 {
-                    object value = primaryKey.GetValue(row);
-                    // Identity columns are allowed to be optional, so just skip the key if it doesn't exist
-                    if (value == null)
+                    // SQL Server allows 900 bytes per primary key, so use that as a baseline
+                    var combinedPrimaryKey = new StringBuilder(900 * table.PrimaryKeys.Count());
+
+                    // Look up primary key of T. Because we're going in the same order of fields every time,
+                    // we can assume that if two rows with the same primary key are in the list, they will collide
+                    foreach (PropertyInfo primaryKey in table.PrimaryKeys)
+                    {
+                        object value = primaryKey.GetValue(row);
+                        // Identity columns are allowed to be optional, so just skip the key if it doesn't exist
+                        if (value == null)
+                        {
+                            continue;
+                        }
+                        combinedPrimaryKey.Append(value.ToString());
+                    }
+                    // If we have already seen this unique primary key, skip this update
+                    if (uniqueUpdatedPrimaryKeys.Add(combinedPrimaryKey.ToString()))
                     {
-                        continue;
+                        // This is the first time we've seen this particular PK. Add this row to the upsert query.
+                        rowsToUpsert.Add(row);
                     }
-                    combinedPrimaryKey.Append(value.ToString());
                 }
-
-                // If we have already seen this unique primary key, skip this update
-                if (uniqueUpdatedPrimaryKeys.Add(combinedPrimaryKey.ToString()))
+                else
                 {
-                    // This is the first time we've seen this particular PK. Add this row to the upsert query.
+                    // ToDo: add check for duplicate primary keys once we find a way to get primary keys.
```

Comment:
As a note - I'm not really sure why this is something you can't do now. Primary keys themselves are a part of the  SQL table so have nothing to do with the item type. 

The only tricky thing is to figure out how to get the values from a JObject - currently this looks like it's storing the MemberInfo from the POCOs and then using those to get the values at runtime for each item/row. But that won't work for JObject - so there will need to be some slight reworking of that class - probably instead having two different implementations of TableInformation - one for JObject and one for the other generic types.

Reply:
yes, I'll work on that after I get this in.